NeoML/src/Dnn/Layers/CrfLayer.cpp

```diff
@@ -123,49 +123,44 @@ bool CCrfCalculationLayer::isFirstStep() const
 
 void CCrfCalculationLayer::RunOnce()
 {
-	int batchWidth = inputBlobs[I_ClassLogProb]->GetBatchWidth();
-	int numberOfClasses = inputBlobs[I_ClassLogProb]->GetObjectSize();
-
 	// The unary estimates of the current elements (the fully-connected layer output)
 	CConstFloatHandle currentProbabilities = inputBlobs[I_ClassLogProb]->GetData();
-	// The vector of partial function value on the previous step (alpha in the forward-backward algorithm, forward pass)
-	// Contains zeros for the first step
-	CConstFloatHandle sequenceProbabilities = inputBlobs[I_ClassSeqLogProb]->GetData();
-	tempSumBlob->Clear();
-	CFloatHandle tempSum = tempSumBlob->GetData();
-	if( !isFirstStep() ) {
+
+	if( isFirstStep() ) {
+		// On the first step all we have to do is initialize the O_ClassSeqLogProb output with current element probabilities
+		MathEngine().VectorCopy( outputBlobs[O_ClassSeqLogProb]->GetData(), currentProbabilities,
+			outputBlobs[O_ClassSeqLogProb]->GetDataSize() );
+	} else {
+		int batchWidth = inputBlobs[I_ClassLogProb]->GetBatchWidth();
+		int numberOfClasses = inputBlobs[I_ClassLogProb]->GetObjectSize();
+
+		// The vector of partial function value on the previous step (alpha in the forward-backward algorithm, forward pass)
+		CConstFloatHandle sequenceProbabilities = inputBlobs[I_ClassSeqLogProb]->GetData();
+
+		tempSumBlob->Clear();
+		CFloatHandle tempSum = tempSumBlob->GetData();
+
 		// Replicate the transitions matrix batchWidth times (manual broadcast)
 		// Add the binary estimates of transition from any previous step to the current
-		// (adding by the previous steps, row by row)
 		MathEngine().AddVectorToMatrixRows(1, tempSum, tempSum, batchWidth,
 			numberOfClasses * numberOfClasses, Transitions()->GetData());
 		// Add the partial function values calculated above
-		// Adding row by row except for the first step when they are equal to 0
 		MathEngine().AddVectorToMatrixRows(batchWidth, tempSum, tempSum,
 			numberOfClasses, numberOfClasses, sequenceProbabilities);
-	}
 
-	if(!IsLearningPerformed()) {
-		// Add the estimates (logits) of the current element classes
-		MathEngine().AddVectorToMatrixColumns(tempSum, tempSum,
```

Comment:
Why this call was removed?

Reply:
It's not removed, it's move to the line 162. As you see, the following argmax is done over rows. So adding a vector `V` to columns of a matrix `M` doesn't affect the result: `argmax(M + V) = argmax(M)`. `O_ClassSeqLogProb` remains correct as well (the max is done over rows): `max(M + V) = V + max(M)`. It's a little optimization, instead of doing broadcast over columns here, we add vector `currentProbabilites` in the end, and get the same result.